File: apps/app-frontend/src/helpers/curseforge.ts
import { invoke } from '@tauri-apps/api/core'

/** Commands for the CurseForge tab (CurseForge mods and modpacks, Feed the Beast modpacks). */

export type CurseForgeClass = 'mod' | 'modpack'
export type CurseForgeSort = 'popularity' | 'updated' | 'downloads' | 'name'
export type CurseForgeLoader = 'forge' | 'neoforge' | 'fabric' | 'quilt'

export type CurseForgeSearchQuery = {
	class: CurseForgeClass
	query?: string | null
	game_version?: string | null
	loader?: CurseForgeLoader | null
	sort: CurseForgeSort
	index?: number
	page_size?: number
}

export type CurseForgeProject = {
	id: number
	name: string
	slug: string
	summary: string
	authors: string[]
	downloads: number
	icon_url: string | null
	website_url: string | null
	updated: string | null
	allow_distribution: boolean
	game_versions: string[]
	loaders: string[]
}

export type CurseForgeSearchResults = {
	projects: CurseForgeProject[]
	index: number
	page_size: number
	total: number
}

export type CurseForgeFile = {
	id: number
	project_id: number
	name: string
	file_name: string
	release_type: 'release' | 'beta' | 'alpha'
	date: string
	size: number
	game_versions: string[]
	loaders: string[]
	downloadable: boolean
	page_url: string
}

export type CurseForgeFiles = {
	files: CurseForgeFile[]
	index: number
	total: number
}

export type ManualDownload = {
	name: string
	url: string
	folder: string
}

export type CurseForgeModInstall = {
	installed: string[]
	manual_downloads: ManualDownload[]
	missing_dependencies: string[]
}

export type PackInstallReport = {
	job: { job_id: string; instance_id: string | null }
	instance_name: string
	manual_downloads: ManualDownload[]
}

export type FtbVersion = {
	id: number
	name: string
	release_type: string
	updated: number
	game_version: string | null
	loader: string | null
	loader_version: string | null
}

export type FtbPack = {
	id: number
	name: string
	summary: string
	icon_url: string | null
	authors: string[]
	installs: number
	plays: number
	updated: number
	tags: string[]
	website_url: string
	versions: FtbVersion[]
}

export type FtbSearchQuery = {
	query?: string | null
	game_version?: string | null
	loader?: string | null
	limit?: number
}

/** Whether this build has a CurseForge API key. Feed the Beast works without one. */
export async function curseforge_status(): Promise<{ available: boolean }> {
	return await invoke('plugin:curseforge|curseforge_status')
}

export async function curseforge_search(
	query: CurseForgeSearchQuery,
): Promise<CurseForgeSearchResults> {
	return await invoke('plugin:curseforge|curseforge_search', { query })
}

export async function curseforge_files(
	projectId: number,
	gameVersion?: string | null,
	loader?: string | null,
	index?: number,
): Promise<CurseForgeFiles> {
	return await invoke('plugin:curseforge|curseforge_files', {
		projectId,
		gameVersion: gameVersion || null,
		loader: loader || null,
		index: index ?? null,
	})
}

/** Installs a mod (the given file, or the newest that fits) and its required dependencies. */
export async function curseforge_install_mod(
	instanceId: string,
	projectId: number,
	fileId?: number | null,
): Promise<CurseForgeModInstall> {
	return await invoke('plugin:curseforge|curseforge_install_mod', {
		instanceId,
		projectId,
		fileId: fileId ?? null,
	})
}

/** Starts installing a CurseForge modpack as a new instance. */
export async function curseforge_install_modpack(
	projectId: number,
	fileId?: number | null,
): Promise<PackInstallReport> {
	return await invoke('plugin:curseforge|curseforge_install_modpack', {
		projectId,
		fileId: fileId ?? null,
	})
}

export async function ftb_search(query: FtbSearchQuery): Promise<FtbPack[]> {
	return await invoke('plugin:curseforge|ftb_search', { query })
}

/** Starts installing a Feed the Beast pack version as a new instance. */
export async function ftb_install_pack(
	packId: number,
	versionId: number,
): Promise<PackInstallReport> {
	return await invoke('plugin:curseforge|ftb_install_pack', { packId, versionId })
}

File: apps/app/src/api/curseforge.rs
//! Commands for Threadrinth's CurseForge tab: CurseForge mods and modpacks
//! and Feed the Beast modpacks. Kept apart from the upstream plugins so
//! Modrinth updates merge cleanly.

use crate::api::Result;
use serde::Serialize;
use tauri::Runtime;
use theseus::curseforge::{
    CurseForgeFiles, CurseForgeModInstall, CurseForgeProject,
    CurseForgeSearchQuery, CurseForgeSearchResults, PackInstallReport,
};
use theseus::ftb::{FtbPack, FtbSearchQuery};

pub fn init<R: Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("curseforge")
        .invoke_handler(tauri::generate_handler![
            curseforge_status,
            curseforge_search,
            curseforge_project,
            curseforge_files,
            curseforge_install_mod,
            curseforge_install_modpack,
            ftb_search,
            ftb_pack,
            ftb_install_pack,
        ])
        .build()
}

#[derive(Serialize)]
pub struct CurseForgeStatus {
    /// Whether this build has a CurseForge API key.
    pub available: bool,
}

#[tauri::command]
pub fn curseforge_status() -> CurseForgeStatus {
    CurseForgeStatus {
        available: theseus::curseforge::is_available(),
    }
}

#[tauri::command]
pub async fn curseforge_search(
    query: CurseForgeSearchQuery,
) -> Result<CurseForgeSearchResults> {
    Ok(theseus::curseforge::search(query).await?)
}

#[tauri::command]
pub async fn curseforge_project(project_id: u32) -> Result<CurseForgeProject> {
    Ok(theseus::curseforge::get_project(project_id).await?)
}

#[tauri::command]
pub async fn curseforge_files(
    project_id: u32,
    game_version: Option<String>,
    loader: Option<String>,
    index: Option<u32>,
) -> Result<CurseForgeFiles> {
    Ok(theseus::curseforge::get_files(
        project_id,
        game_version,
        loader,
        index.unwrap_or(0),
    )
    .await?)
}

/// Installs a mod (the given file, or the newest one that fits) and its
/// required dependencies into an instance.
#[tauri::command]
pub async fn curseforge_install_mod(
    instance_id: &str,
    project_id: u32,
    file_id: Option<u32>,
) -> Result<CurseForgeModInstall> {
    Ok(
        theseus::curseforge::install_mod(instance_id, project_id, file_id)
            .await?,
    )
}

/// Starts installing a CurseForge modpack as a new instance.
#[tauri::command]
pub async fn curseforge_install_modpack(
    project_id: u32,
    file_id: Option<u32>,
) -> Result<PackInstallReport> {
    Ok(theseus::curseforge::install_modpack(project_id, file_id).await?)
}

#[tauri::command]
pub async fn ftb_search(query: FtbSearchQuery) -> Result<Vec<FtbPack>> {
    Ok(theseus::ftb::search(query).await?)
}

#[tauri::command]
pub async fn ftb_pack(pack_id: u32) -> Result<FtbPack> {
    Ok(theseus::ftb::get_pack(pack_id).await?)
}

/// Starts installing a Feed the Beast pack version as a new instance.
#[tauri::command]
pub async fn ftb_install_pack(
    pack_id: u32,
    version_id: u32,
) -> Result<PackInstallReport> {
    Ok(theseus::ftb::install_pack(pack_id, version_id).await?)
}
